src/session_manager.rs

```diff
@@ -0,0 +1,1180 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+use tdgrand::enums::{self, AuthorizationState, Update};
+use tdgrand::functions;
+use tdgrand::types::{self, UpdateAuthorizationState};
+use tokio::fs::{self, File};
+use tokio::io::AsyncReadExt;
+use tokio_stream::wrappers::ReadDirStream;
+
+use crate::utils::{
+    data_dir, do_async, send_tdlib_parameters, session_info_path, write_session_info_to_file,
+};
+use crate::Session;
+use crate::APPLICATION_OPTS;
+use crate::RUNTIME;
+
+const LEAGACY_DB_NAME: &str = "db0";
+
+/// Struct for storing information about a td client.
+#[derive(Clone, Debug)]
+pub enum ClientInfo {
+    /// The client is currently in the authorization state. Every client, even those that were
+    /// logged in during a previous run of the application will need to go through this state
+    /// again.
+    Auth {
+        /// Whether the client needs to authorize. If a client had the `AuthorizationState::Ready`
+        /// state already from a previous run of the application this will be set to `false`.
+        /// In this case it is expected that the client will always get `AuthorizationState::Ready`
+        /// again.
+        /// We use this information to bypass the login process (phone number -> auth code -> ...
+        /// -> ready) and to just wait for the `AuthorizationState::Ready` update.
+        needs_auth: bool,
+        /// States whether the client uses a test data center.
+        use_test_dc: bool,
+        /// The directory name of the database for this client.
+        database_dir_base_name: String,
+        /// States whether this client is currently the active client that is shown to the user.
+        /// We'll switch to this in the session stack after startup.
+        is_active: bool,
+    },
+    /// The client is logged and has a `Session`.
+    LoggedIn(
+        /// The `Session` of this client.
+        Session,
+    ),
+}
+impl ClientInfo {
+    fn database_dir_base_name(&self) -> &str {
+        match self {
+            Self::Auth {
+                database_dir_base_name,
+                ..
+            } => database_dir_base_name,
+            Self::LoggedIn(session) => session.database_dir_base_name(),
+        }
+    }
+}
+
+mod imp {
+    use super::*;
+
+    use std::cell::RefCell;
+    use std::collections::HashMap;
+
+    use crate::Login;
+
+    #[derive(Debug, Default, CompositeTemplate)]
+    #[template(resource = "/com/github/melix99/telegrand/ui/session-manager.ui")]
+    pub struct SessionManager {
+        /// The order of the recently used sessions. The string stored in the `Vec` represents
+        /// the sessions's database directory name.
+        pub recently_used_sessions: RefCell<Vec<String>>,
+        pub clients: RefCell<HashMap<i32, ClientInfo>>,
+        #[template_child]
+        pub main_stack: TemplateChild<gtk::Stack>,
+        #[template_child]
+        pub login: TemplateChild<Login>,
+        #[template_child]
+        pub sessions: TemplateChild<gtk::Stack>,
```

Comment:
What about storing sessions in a HashMap? It will make the transition to a "single widget tree" less painful, and it will probably make this code a bit better. You could use an AdwBin to decide which session to show using `set_child()`. If it can help, you can also store the active session in a separate parameter (I will need to do that either way).

Reply:
> EDIT: another thing to consider (if useful) is adding an id parameter to the session, maybe storing just the numeric part of the db name.

I see two problems here.

The first is that the  `db<t>` scheme is not a hard constraint for a name of a database. It's only the scheme used to create new databases in the application. This makes the whole thing much more flexible and also easier to implement.  There would be nothing against the user renaming a database according to his wishes.

The second thing is that even within the application the scheme can be broken.
https://github.com/melix99/telegrand/blob/ea93128e1b8cf472615e83992c66aacb51eea72d/src/session_manager.rs#L976-L998
Admittedly, this is extremely unlikely (e.g., misconfigured clock), but one should be prepared for anything. :-P

> What about storing sessions in a HashMap (or Vec)? It will make the transition to a "single widget tree" less painful, and it will probably make this code a bit better. You could use an AdwBin to decide which session to show using `set_child()`. If it can help, you can also store the active session in a separate parameter (I will need to do that either way).

This would open a can of worms for this PR, as the `SelectionModel` of the `Stack` is used in several places. So I would be very grateful to you to do that in another PR.

On a side note, I think we should definitely start a discussion about single widget tree vs multi widget tree.

I've been second-guessing myself the last few days/weeks about whether a single widget tree is the way to go. If I see it correctly, this approach would actually only have disadvantages.

For example, `ChatHistory` is based on a single widget tree and to be honest, the UX is not very great: there is a noticeable delay (stuttering) when switching chats. Also, visual status is lost. When you jump back to a chat, you are no longer at the same place in the history. Bad UX is my main gripe.

I also think that the single widget tree is technically the more error-prone approach. You always have to watch that bindings have to be re-set, bindings have to be removed, etc.

Also, I think it helps not to deviate too much from `Fractal`. Thus, improvements from Fractal can be adopted much more easily by us.  

Lastly, I think the single widget tree is the more inflexible one. For example, if we ever decided to display sessions in tabs instead of a ListView, we would need to change almost nothing.

I wonder what the advantage of a single widget tree is. Maybe less resource consumption? If so, I wonder how much it is and if it is worth it given all the disadvantages I see. Of course, we can only find out if we measure it. :-)

Reply:
> I wonder what the advantage of a single widget tree is. Maybe less resource consumption?

Yes, but that's a bit more complicated than that. Having multiple listviews existing at the same time means that we will have hundreds of widget (the max count before they will start recycling is 200 for now) for every open chat. This widgets will also probably have a signal or expression needed for updating the content for when the messaged are edited. We'd need a system to avoid having too many signals existing at the same time, because they eat resources. Also, we'd need to also at least destroy the media widgets, because they will eat a lot of ram.

> For example, ChatHistory is based on a single widget tree and to be honest, the UX is not very great: there is a noticeable delay (stuttering) when switching chats. Also, visual status is lost. When you jump back to a chat, you are no longer at the same place in the history. Bad UX is my main gripe.

I understand and agree with your points, currently having a single recyclable list for every single chat means that the UX is really bad. But quite honestly I think these problems should be solved by gtk itself by improving their recycling widgets (for example allowing the list to scroll at a given position, also taking in consideration the recycling, also something like default-to-bottom position of the items, etc.). But that means ofc that we should also contribute to gtk, which would be ok, but the listview.c and listbase.c are not exactly small and easy to contribute to.

Now, I agree that we should at least measure how things will go with storing all the widgets at the same time and then evaluate if this is actionable or not. Then in the future, when (hopefully) the list widgets will get better, we could go back with having a single widget tree. I'll open a separate PR for this (I'll probably test the `ChatHistory` first).

For now I think this is resolvable anyway, the most important part for this PR is having a stable "db directory system".